// smart-http/.eslintrc.js
module.exports = {
  env: {
    es6: true,
    node: true,
  },
  rules: {
    'arrow-parens': ['error', 'always'],
    'comma-dangle': ['error', 'always-multiline'],
    indent: ['error', 2],
    'max-len': [
      'error',
      {
        code: 160,
        comments: 160,
        ignoreUrls: true,
      },
    ],
    'max-lines': [
      'error',
      {
        max: 500,
        skipBlankLines: true,
        skipComments: true,
      },
    ],
    'newline-after-var': ['error', 'always'],
    'newline-before-return': 'error',
    'no-debugger': 'error',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-empty': 'error',
    'no-empty-pattern': 'error',
    'no-eval': 'error',
    'no-extra-semi': 'error',
    'no-irregular-whitespace': [
      'error',
      {
        skipComments: true,
        skipRegExps: true,
        skipTemplates: true,
      },
    ],
    'no-label-var': 'error',
    'no-lonely-if': 'error',
    'no-mixed-requires': 'error',
    'no-multi-str': 'error',
    'no-multiple-empty-lines': [
      'error',
      {
        max: 1,
        maxEOF: 0,
      },
    ],
    'no-path-concat': 'error',
    'no-process-exit': 'error',
    'no-restricted-globals': 'error',
    'no-tabs': 'error',
    'no-trailing-spaces': 'error',
    'no-unexpected-multiline': 'error',
    'no-unneeded-ternary': 'error',
    'no-unreachable': 'error',
    'no-unsafe-finally': 'error',
    'no-useless-concat': 'error',
    'no-useless-constructor': 'error',
    'no-useless-return': 'error',
    'no-var': 'error',
    'object-curly-spacing': [
      'error',
      'always',
      {
        arraysInObjects: true,
        objectsInObjects: true,
      },
    ],
    'prefer-const': 'error',
    'prefer-destructuring': 'error',
    'prefer-spread': 'error',
    quotes: ['error', 'single'],
  },
};
