// smart-auth-middleware/lib/jwt-verify.js
const jwt = require('express-jwt');
const jwksClient = require('jwks-rsa');
const fs = require('fs');

/**
 * define our authentication middleware
 * see https://github.com/auth0/node-jwks-rsa#caching
 * see https://github.com/auth0/node-jwks-rsa#rate-limiting
 * validate the audience & issuer from received token vs JWKS endpoint
 * originalUrl only contains the url endpoint like ('/ping'|'/healthcheck') not the full fledge url endpoint like- http://abc.com:80/ping
 */
const jwtVerify = ({
  IDENTITY_SERVICE_URL, ISSUER = 'fininfinity technology private limited', JWKSURL, AUDIENCE, PUBLIC_KEY_PATH,
}) =>
  jwt({
    secret: PUBLIC_KEY_PATH ? fs.readFileSync(PUBLIC_KEY_PATH, 'utf8') : jwksClient.expressJwtSecret({
      cache: true,
      rateLimit: true,
      jwksRequestsPerMinute: 2,
      jwksUri: JWKSURL || `${IDENTITY_SERVICE_URL}/.well-known/jwks`,
    }),
    audience: AUDIENCE || 'platform',
    issuer: ISSUER || IDENTITY_SERVICE_URL,
    algorithms: [ 'RS256' ],
  });

module.exports = jwtVerify;
